### ISSUE TITLE
Legion/Realm: interface for CUDA constants

### ISSUE THREAD

**Issue by mariodirenzo** [StanfordLegion/legion#964]
To my understanding the responsibility of managing the CUDA constant variables is currently left to users who need to call `cudaMemcpyToSymbol` at the right time, on the right device, and in the right cuda stream in order to ensure that the constant is well defined when the task is executed. Considering the execution model of Legion, this might be very difficult and might generate races between tasks that set the same constant at the same time. Would it be possible to create a way to register the CUDA constants (with their expected values) and insert them in the task requirements?

**Comment by streichler** [comment]
Will the values you want to have copied into the constant variables be produced by another task, or are they compile-time constants?  How often do you expect to change the contents of the constant variables during execution?

**Comment by mariodirenzo** [comment]
Right now they are constants coming from the input to the code, they are set at the beginning of the run and they stay the same. In the future, they might change every once in a while during the execution

**Comment by lightsighter** [comment]
Note @streichler if you want to make a new kind of "constant" memory for each GPU and allow the mapping of constant instances, that would be fine with me.

**Comment by elliottslaughter** [comment]
Can someone tell me what the impact of this issue is? Does it prevent constant memory from being used in C++ CUDA tasks at all? Is there something @mariodirenzo is doing that's unusual and somehow different from what the CUDA compiler normally does?

Is this a dependency for #638?

**Comment by streichler** [comment]
> Is this a dependency for #638?

No, the regent codegen can generate an explicit call to `cudaMemcpyToSymbol` before it launches a kernel and the right thing will happen.  (This optimization is used in Kokkos as well).

My interpretation of this request is to be able to map long-lived instances in constant memory so that they don't need to be downloaded for every kernel launch.

**Comment by lightsighter** [comment]
> My interpretation of this request is to be able to map long-lived instances in constant memory so that they don't need to be downloaded for every kernel launch.

That's correct, this would be for mapping actual Legion regions into constant cache memory so that multiple kernels across different tasks could access them.

**Comment by mariodirenzo** [comment]
> Is there something @mariodirenzo is doing that's unusual and somehow different from what the CUDA compiler normally does?

From what I know (and I do not know if there is any workaround), CUDA constants are defined at the global scope and will therefore live for the entire duration of the execution. Considering that the amount of constant cache, the idea is that different tasks might use the same CUDA constant instance for their kernels. If all the tasks are designed to read the same value from the constant instance, having multiple `cudaMemcpyToSymbol` in each task cannot create side effects. If the tasks are supposed to read different values from the same constant instance, the tasks scheduled for the same GPU might create a race between their kernels and the `cudaMemcpyToSymbol` that they issue.
